packages/common/observable-object/src/access-observer.ts

```diff
@@ -0,0 +1,127 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { EventSubscriptions } from '@dxos/async';
+
+import { ObservableObject, subscribe } from './observable-object';
+
+export const ACCESS_OBSERVER_STACK: AccessObserver[] = [];
+
+// TODO(burdon): Document or remove. Generic filter function?
+export type SelectionFn = never;
+
+export type Selection = ObservableObject | SelectionFn | Selection[] | undefined | null | false;
+
+// TODO(dmaretskyi): Convert to class.
+export interface SubscriptionHandle {
+  update: (selection: Selection) => SubscriptionHandle;
+  subscribed: boolean;
+  unsubscribe: () => void;
+  selectedIds: Set<string>;
+}
+
+export type UpdateInfo = {
+  // TODO(dmaretskyi): Include metadata about the update.
+  // updatedIds: string[];
+};
+
+export const createAccessObserver = () => {
+  const observer = new AccessObserver(() => {
+    ACCESS_OBSERVER_STACK.splice(ACCESS_OBSERVER_STACK.indexOf(observer), 1);
+  });
+  ACCESS_OBSERVER_STACK.push(observer);
+  return observer;
+};
+
+export const logObjectAccess = (obj: ObservableObject) => {
+  ACCESS_OBSERVER_STACK.at(-1)?.accessed.add(obj);
+  // TODO(wittjosiah): Print a helpful warning if we're accessing data without observing it.
+  //   Needs to only print once per component.
+  // if (this._accessObserverStack.length === 0) {
+  //   const currentComponent = getCurrentReactComponent();
+  //   if (currentComponent) {
+  //     log.warn(
+  //       `Warning: Data access in a React component without \`observer\`. Component will not update correctly.\n  at ${currentComponent.fileName}:${currentComponent.lineNumber}`
+  //     );
+  //   }
+  // }
+};
+
+/**
+ * Subscribe to database updates.
+ * Calls the callback when any object from the selection changes.
+ * Calls the callback when the selection changes.
+ * Always calls the callback on the first `selection.update` call.
+ */
+// TODO(burdon): Add filter?
+// TODO(burdon): Immediately trigger callback.
+export const createSubscription = (
+  observer: AccessObserver,
+  onUpdate: (info: UpdateInfo) => void,
+): SubscriptionHandle => {
+  let subscribed = true;
+
+  const subscriptions = Array.from(observer.accessed.values()).reduce((acc, obj) => {
+    acc.add(obj[subscribe](onUpdate));
+
+    return acc;
+  }, new EventSubscriptions());
+
+  let firstUpdate = true;
+
+  const handle = {
```

Comment:
might be a good time to turn this into a class

Reply:
what would be the goal of making it a class?

Reply:
its functionally already a class: it has fields and methods. Just uses wierd syntax right now